---
toc: false
title: Linnorm User Manual
author:
  Shun H. Yip^1,2,3^, Panwen Wang^3^, Jean-Pierre Kocher^3^, Pak Chung Sham^1,4,5^, Junwen Wang^3,6^
output:
 BiocStyle::pdf_document
vignette: >
 %\VignetteIndexEntry{Linnorm User Manual}
 %\VignetteEngine{knitr::rmarkdown}
 %\VignetteEncoding{UTF-8}
---




\begin{center}
\date{October 07, 2016}


\small{$^{1}$ Centre for Genomic Sciences, LKS Faculty of Medicine,}\protect \\ 
\small{The University of Hong Kong, Hong Kong SAR, China;}\protect \\
\small{$^{2}$ School of Biomedical Sciences, LKS Faculty of Medicine,}\protect \\
\small{The University of Hong Kong, Hong Kong SAR, China;}\protect \\
\small{$^{3}$ Department of Health Sciences Research and Center for}\protect \\
\small{Individualized Medicine, Mayo Clinic, Scottsdale,}\protect \\
\small{AZ, 85259, USA;}\protect \\
\small{$^{4}$ Department of Psychiatry, LKS Faculty of Medicine,}\protect  \\
\small{The University of Hong Kong, Hong Kong SAR, China;}\protect \\
\small{$^{5}$ State Key Laboratory in Cognitive and Brain Sciences,}\protect  \\
\small{The University of Hong Kong, Hong Kong SAR, China;}\protect \\
\small{$^{6}$ Department of Biomedical Informatics, Arizona State}\protect  \\
\small{University, Scottsdale, AZ, 85259, USA.}\protect \\


\end{center}



\bigskip

\begin{abstract}
 Linnorm is an R package for the analysis of RNA-seq, scRNA-seq, ChIP-seq count data or any large scale count data. Its main function is to normalize and transform such datasets for parametric tests. Various analysis pipelines are also implemented for users' convenience, including using \Biocpkg{limma} for differential expression analysis or differential peak detection\footnote{The Linnorm-limma pipeline is implemented as the "Linnorm.limma" function. Please cite both Linnorm and limma if you use this function for publication.}, calculating correlation coefficient for gene correlation study\footnote{Implemented as the Linnorm.Cor function.}, subpopulation analysis pipeline with PCA and k-means clustering\footnote{Implemented as the Linnorm.PCA function.}, highly variable gene discovery\footnote{Implemented as the Linnorm.HVar function.} and hierarchical clustering analysis\footnote{Implemented as the Linnorm.HClust function.}. Linnorm can work with raw count, CPM, RPKM, FPKM and TPM and is compatible with data generated from simple count algorithms\footnote{Such as \href{http://www-huber.embl.de/HTSeq/doc/overview.html}{HTSeq}, \Biocpkg{Rsubread} and etc} and supervised learning algorithms\footnote{Such as \href{http://cole-trapnell-lab.github.io/cufflinks/}{Cufflinks}, \href{http://bio.math.berkeley.edu/eXpress/index.html}{eXpress}, \href{https://pachterlab.github.io/kallisto/}{Kallisto}, \href{http://deweylab.github.io/RSEM/}{RSEM}, \href{http://www.cs.cmu.edu/~ckingsf/software/sailfish/}{Sailfish}, and etc}.
 
 Additionally, Linnorm provides the RnaXSim function for the simulation of RNA-seq raw counts for the evaluation of differential expression analysis methods. RnaXSim can simulate RNA-seq dataset in Gamma, Log Normal, Negative Binomial or Poisson distribution.
\end{abstract}

\newpage

\setcounter{tocdepth}{4}
\tableofcontents

\newpage
#Introduction

Linnorm is a count data transformation method. Linnorm transforms the dataset toward both homoscedasticity and normality; and it only has one transformation parameter for the whole dataset. This approach ensures that the resulting dataset can better satisfy the assumptions made by the linear models, which is used by \Biocpkg{limma}. Furthermore, since Linnorm is transforming the dataset toward homoscedasticity and normality, it performs well with datasets which may be deviated from the negative binomial distribution, such as those generated by supervised learning based gene quantification software. Additionally, since homoscedasticity and normality are also assumed by many other parametric tests, such as PCA and the Pearson correlation coefficient, Linnorm transformed datasets are not restricted to differential expression analysis.

The Linnorm R package contains a variety of functions for RNA-seq data analysis. It has three functions.

* RNA-seq Expression Normalization/Transformation \( Linnorm \)
	* Output transformed data matrix for analysis
    * Gene Correlation Analysis
* The Linnorm-limma pipeline \( Linnorm.limma \)
	* Differential expression analysis
	* Differential peak detection
* PCA k-means clustering \( Linnorm.PCA \)
	* Single cell RNA-seq subpopulation analysis
* Highly variable gene discovery \( Linnorm.HVar \)
* Gene correlation analysis \( Linnorm.Cor \)
* Hierarchical clustering analysis \( Linnorm.HClust \)
* RNA-seq Raw Count Simulation \( RnaXSim \)

##Linnorm

###Datatypes and Input Format

Linnorm accepts any RNA-seq Expression data, including but not limited to

* Raw Count \( RNA-seq or ChIP-seq \)
* Count per Million       \( CPM \)
* Reads per Kilobase per Million reads sequenced       \( RPKM \)
* expected Fragments Per Kilobase of transcript per Million fragments sequenced       \( FPKM \)
* Transcripts per Million       \( TPM \)

We suggest RPKM, FPKM or TPM for most purposes.

Linnorm accepts matrix as its data type. Data frames are also accepted and will be automatically converted into the matrix format before analysis. Each column in the matrix should be a sample or replicate. Each row should be a Gene/Exon/Isoform/etc.

Example:

|     |Sample 1|Sample 2|Sample 3|...  |
|:---:|:------:|:------:|:------:|:----|
|Gene1|1       |2       |1       |...  |
|Gene2|3       |0       |4       |...  |
|Gene3|10.87   |11.56   |12.98   |...  |
|...  |...     |...     |...     |...  |
|...  |...     |...     |...     |...  |

Please note that undefined values such as NA, NaN, INF, etc are **NOT** supported.

By using the argument, "input = "Linnorm"", functions provided by the Linnorm package can also accept Linnorm transformed datasets as input.

##RnaXSim

The Linnorm package provides the RnaXSim function for the simulation of RNA-seq expression data given a distribution. Supported distributions include:

* Gamma Distribution
* Log Normal Distribution
* Negative Binomial Distribution
* Poisson Distribution

###Inputs

Acceptable input format, expression types and data types are the same as Linnorm's section.

Each sample in the data matrix is assumed to be replicates of each other. The \Biocpkg{seqc} package is a good source of such datasets.

\newpage
#Examples with Source Codes

##Linnorm

###Linnorm-limma Differential Expression Analysis

* limma package
	- \Biocpkg{limma} is imported with Linnorm. Please cite both Linnorm and limma if you use the Linnorm.limma function for differential expression analysis for publication.

####Obtain example data


1. Get 20 samples of RNA-seq data. These 20 samples are paired tumor and adjacent normal tissue samples from 10 individuals from TCGA LIHC dataset.

```{r, echo=TRUE}
library(Linnorm)
data(LIHC)
datamatrix <- LIHC
```

####Analysis procedure


The Linnorm-limma pipeline only consists of two steps.

1. Create limma design matrix
```{r, echo=TRUE}
#10 samples for condition 1 and 10 samples for condition 2.
#You might need to edit this line
design <- c(rep(1,10),rep(2,10))
#Just copy these lines to R
design <- model.matrix(~ 0+factor(design))
colnames(design) <- c("group1", "group2")
rownames(design) <- colnames(datamatrix)
```


2. Linnorm-limma Differential Expression Analysis

a. Basic Differential Expression Analysis. (Follow this if you are not sure what to do.)
```{r, echo=TRUE}
library(Linnorm)
#The Linnorm-limma pipeline only consists of one line.
DEG_Results <- Linnorm.limma(datamatrix,design)
#The DEG_Results matrix contains your DEG analysis results.
```


b. Advanced: to output both DEG analysis results and the transformed matrix for further analysis
```{r, echo=TRUE}
library(Linnorm)
#Just add output="Both" into the argument list
BothResults <- Linnorm.limma(datamatrix,design,output="Both")

#To separate results into two matrices:
DEG_Results <- BothResults$DEResults
TransformedMatrix <- BothResults$Linnorm
#The DEG_Results matrix now contains DEG analysis results.
#The TransformedMatrix matrix now contains a Linnorm
#Transformed dataset.
```



####Print out the most significant genes

1. Write out the results to a tab delimited file.
```{r}
write.table(DEG_Results, "DEG_Results.txt", quote=FALSE, sep="\t",
col.names=TRUE,row.names=TRUE)
```

2. Print out the most significant 10 genes.
```{r, echo=TRUE}
Genes10 <- DEG_Results[order(DEG_Results[,"adj.P.Val"]),][1:10,]
#Users can print the gene list by the following command:
#print(Genes10)

#logFC: log 2 fold change of the gene.
#XPM: If input is raw count or CPM, XPM is CPM.
#	If input is RPKM, FPKM or TPM, XPM is TPM.
#t: moderated t statistic.
#P.Value: p value.
#adj.P.Val: Adjusted p value. This is also called False Discovery Rate or q value.}
#B: log odds that the feature is differential.
```
```{r,kable1, echo=FALSE}
library(knitr)
kable(Genes10, digits=4)
```


#### Volcano Plot
1. Remove Genes which fold changes are INF. You can skip this if
there are no INF values in the fold change column.
```{r, echo=TRUE}
NoINF <- DEG_Results[which(!is.infinite(DEG_Results[,1])),]
```

2. Record significant genes for coloring
```{r, echo=TRUE}
SignificantGenes <- NoINF[NoINF[,5] <= 0.05,1]
```

3. Draw volcano plot with Log q values. 
Green dots are non-significant, red dots are significant.
```{r, echo=TRUE, fig.height=6, fig.width=6}
plot(x=NoINF[,1], y=NoINF[,5], col=ifelse(NoINF[,1] %in%
SignificantGenes, "red", "green"),log="y", ylim = 
rev(range(NoINF[,5])), main="Volcano Plot", 
xlab="log2 Fold Change", ylab="q values", cex = 0.5)
```



### Single cell RNA-seq DEG Analysis

By default, Linnorm filters gene with excess zero counts for data modeling, which is optimized for RNA-seq data. Since single cell RNA-seq data often contain a larger amount of zeroes, it is suggested to adjust the minZeroPortion argument during analysis

In this section, we use Islam et al. (2011)'s single cell RNA-seq dataset to perform DEG analysis. In this analysis, we are using 48 mouse embryonic stem cells and 44 mouse embryonic fibroblasts.

Read data:
```{r, echo=TRUE}
library(Linnorm)
data(Islam2011)
IslamData <- Islam2011[,1:92]
```

1. Create limma design matrix
```{r, echo=TRUE}
#48 samples for condition 1 and 44 samples for condition 2.
#You might need to edit this line
design <- c(rep(1,48),rep(2,44))
#Just copy these lines to R
design <- model.matrix(~ 0+factor(design))
colnames(design) <- c("group1", "group2")
rownames(design) <- colnames(IslamData)

```
2. DEG Analysis
```{r, echo=TRUE}
#Example 1: Filter low count genes
scRNAseqResults <- Linnorm.limma(IslamData,design,minZeroPortion=0.5, keepAll=FALSE)


#Example 2: Do not filter gene list.
scRNAseqResults <- Linnorm.limma(IslamData,design,keepAll=TRUE)
```



\newpage
###PCA K-means Clustering. Cell subpopulation analysis.

In this section, we use Islam et al. (2011)'s single cell RNA-seq dataset to perform subpopluation analysis. The 96 samples are consisted of 48 mouse embryonic stem cells, 44 mouse embryonic fibroblasts and 4 negative controls. We do not use the negative controls here.

1. Read data.
```{r, echo=TRUE}
library(Linnorm)
data(Islam2011)
```

####Simple subpopulation analysis
1. Perform PCA analysis using default configurations. Only keep genes with less than half of the samples being zero.
```{r, echo=TRUE}
PCA.results <- Linnorm.PCA(Islam2011[,1:92])
```

2. Draw PCA k-means clustering plot.
```{r, echo=TRUE, fig.height=6, fig.width=6}
#Here, we can see three clusters.
print(PCA.results$plot$plot)
```


####Analysis with known subpopulations
1. Assign known groups to samples.
```{r, echo=TRUE}
#The first 48 samples belong to mouse embryonic stem cells.
Groups <- rep("ES_Cells",48)
#The next 44 samples are mouse embryonic fibroblasts.
Groups <- c(Groups, rep("EF_Cells",44))
```

2. Perform PCA analysis.
```{r, echo=TRUE}
#Useful arguments:
#Group:
#allows user to provide each sample's information.
#num_center:
#how many clusters are supposed to be there.
#num_PC
#how many principal components should be used in k-means
#clustering.

PCA.results <- Linnorm.PCA(Islam2011[,1:92],
minZeroPortion=0.5, Group=Groups, num_center=3, num_PC=2)
```

3. Draw PCA k-means clustering plot.
```{r, echo=TRUE, fig.height=6, fig.width=6}
#Here, we can see two clusters.
print(PCA.results$plot$plot)
```


\newpage
###Gene Correlation Network Analysis
In this section, we are going to use Islam2011 single cell RNA-seq embryonic stem cell data and perform Gene Correlation Analysis.

1. Obtain data.
```{r, echo=TRUE}
data(Islam2011)

#Obtain embryonic stem cell data
datamatrix <- Islam2011[,1:48]
```

2. Perform analysis.
```{r, echo=TRUE}
#Setting plotNetwork to TRUE will create a figure file in your current directory.
#Setting it to FALSE will still create an igraph object, so that users can plot it
#manually later.
#For this vignette, we will plot it manually instead in step 4.
results <- Linnorm.Cor(datamatrix, plotNetwork=FALSE,
#Edge color when correlation is positive
plot.Pos.cor.col="red",
#Edge color when correlation is negative
plot.Neg.cor.col="green")
```

3. Print out the most significant 10 pairs of genes.
```{r, echo=TRUE}
Genes10 <- results$Results[order(results$Results[,4]),][1:10,]
#Users can print the gene list by the following command:
#print(Genes10)
```
```{r,kable2, echo=FALSE}
library(knitr)
kable(Genes10, digits=4, row.names=FALSE)
```


#### Plot a co-expression network.
We will demonstrate how to plot the figure "networkplot.png" here.

```{r, echo=TRUE, fig.height=6, fig.width=6}
library(igraph)
Thislayout <- layout_with_kk(results$igraph)
plot(results$igraph,
#Vertex size
vertex.size=1.5,
#Vertex color, based on clustering results
vertex.color=results$Cluster$Cluster,
#Vertex frame color
vertex.frame.color="transparent",
#Vertex label color (the gene names)
vertex.label.color="black",
#Vertex label size
vertex.label.cex=0.05,
#This is how much the edges should be curved.
edge.curved=0.1,
#Edge width
edge.width=0.05,
#Use the layout created previously
layout=Thislayout
)
```

#### Identify genes that belong to a cluster.
For example, what are the genes that belong to the same cluster as the Mmp2 gene?
a. Identify the cluster that Mmp2 belongs to.
```{r, echo=TRUE}
TheCluster <- which(results$Cluster[,1] == "Mmp2")
```

b. Obtain the list of genes that belong to this cluster.
```{r, echo=TRUE}
#Index of the genes
ListOfGenes <- which(results$Cluster[,2] == TheCluster)

#Names of the genes
GeneNames <- results$Cluster[ListOfGenes,1]

#Users can write these genes into a file for further analysis.
```

#### Draw a correlation heatmap.
1. Randomly choose 100 genes from clustering results
```{r, echo=TRUE}
top100 <- results$Cluster[sample(1:nrow(results$Cluster), 100),1]
```
2. Extract these 100 genes from the correlation matrix.
```{r, echo=TRUE}
Top100.Cor.Matrix <- results$Cor.Matrix[top100,top100]
```
3. Draw a correlation heatmap.
```{r, echo=TRUE, fig.height=6, fig.width=6}
library(RColorBrewer)
library(gplots)
heatmap.2(as.matrix(Top100.Cor.Matrix),
#Hierarchical clustering on both row and column
Rowv=TRUE, Colv=TRUE,
#Center white color at correlation 0
symbreaks=TRUE,
#Turn off level trace
trace="none",
#x and y axis labels
xlab = 'Genes', ylab = "Genes",
#Turn off density info
density.info='none',
#Control color
key.ylab=NA,
col=colorRampPalette(c("blue", "white", "yellow"))(n = 1000),
lmat=rbind(c(4, 3), c(2, 1)),
#Gene name font size
cexRow=0.3,
cexCol=0.3,
#Margins
margins = c(8, 8)
)
```

###Highly variable gene analysis
In this section, we will perform highly variable gene discovery on the embryonic stem cells from Islam(2011).
1. Obtain data.
```{r, echo=TRUE}
data(Islam2011)

#Identify spike in genes:
SPIKEIN <- rownames(Islam2011)[c(grep("Ppia",rownames(Islam2011)),
grep("H2afz",rownames(Islam2011)),grep("Hprt1",rownames(Islam2011)))]
```

2. Analysis
```{r, echo=TRUE}
#The first 48 columns are the embryonic stem cells.
results <- Linnorm.HVar(Islam2011[,1:48], spikein=SPIKEIN)
```

3. Print out the most significant 10 genes.
```{r, echo=TRUE}
resultsdata <- results$Results
Genes10 <- resultsdata[order(resultsdata[,"q.value"]),][1:10,1:5]
#Users can print the gene list by the following command:
#print(Genes10)
```
```{r,kable3, echo=FALSE}
library(knitr)
kable(Genes10, digits=4)
```

#### Mean vs SD plot highlighting significant genes.
1. Simply print the plot from results.
```{r, echo=TRUE, fig.height=6, fig.width=6}
print(results$plot$plot)
```

### Hierarchical Clustering
In this section, we will perform hierarchical clustering on Islam2011 data.
1. Obtain data.
```{r, echo=TRUE}
data(Islam2011)
```

2. Assign group to samples.
```{r, echo=TRUE}
#48 ESC, 44 EF, and 4 NegCtrl
Group <- c(rep("ESC",48),rep("EF",44),rep("NegCtrl",4))
colnames(Islam2011) <- paste(colnames(Islam2011),Group,sep="_")
```

3. Perform Analysis.
```{r, echo=TRUE}
#Note that there are 3 known clusters.
HClust.Results <- Linnorm.HClust(Islam2011,Group=Group, num_Clust=3, fontsize=1.8)
```

#### Hierarchical Clustering plot
We can simply print the plot out.
```{r, echo=TRUE, fig.height=6, fig.width=6}
print(HClust.Results$plot$plot)
```

###Basic Linnorm Transformation

Here, we will demonstrate how to generate and ouput Linnorm Transformed dataset into a tab delimited file.


1. Linnorm Transformation
```{r, echo=TRUE}
library(Linnorm)
Transformed <- Linnorm(Islam2011, minZeroPortion=0.5)
#minZeroPortion is suggested to be 0.5 for scRNA-seq.
```


2. Write out the results to a tab delimited file.
```{r, echo=TRUE}
#You can use this file with Excel.
write.table(Transformed, "Transformed_Matrix.txt",
quote=FALSE, sep="\t", col.names=TRUE, row.names=TRUE)
```


####Calculate Fold Change


Fold change can be calculated by the Linnorm.limma function. It is included in differential expression analysis results. However, for users who would like to calculate fold change from Linnorm transformed dataset and analyze it. Here is an example.


1. Get 20 samples of TCGA RNA-seq data.
```{r, echo=TRUE}
library(Linnorm)
data(LIHC)
datamatrix <- LIHC
```


2. Linnorm Transformation.
```{r, echo=TRUE}
library(Linnorm)
LNormData <- Linnorm(datamatrix)
```


3. Undo Logarithm from LNormData.
```{r, echo=TRUE}
#Let LNormData be a matrix of Linnorm Transformed dataset. 
Newdata <- exp(LNormData)
```

4. Calculate Fold Change.
```{r, echo=TRUE}
#Now, we can calculate fold changes between
#sample set 1 and sample set 2.
#Index of sample set 1 from LNormData and Newdata:
set1 <- 1:10
#Index of sample set 2 from LNormData and Newdata:
set2 <- 11:20

#Define a function that calculates log 2 fold change:
log2fc <- function(x) {
return(log(mean(x[set1])/mean(x[set2]),2))
}

#Calculate log 2 fold change of each gene in the dataset:
foldchanges <- unlist(apply(Newdata,1,log2fc))

#Put resulting data into a matrix
FCMatrix <- matrix(nrow=length(foldchanges),ncol=1)
rownames(FCMatrix) <- rownames(LNormData)
colnames(FCMatrix) <- c("Log 2 Fold Change")
FCMatrix[,1] <- foldchanges

#Now FCMatrix contains fold change results.
```

5. Draw a probability density plot of the fold changes in the dataset.

```{r, echo=TRUE, fig.height=6, fig.width=6}
Density <- density(foldchanges)
plot(Density$x,Density$y,type="n",xlab="Log 2 Fold Change",
ylab="Probability Density",)
lines(Density$x,Density$y, lwd=1.5, col="blue")
title("Probability Density of Fold Change.\nTCGA Partial LIHC data
Paired Tumor vs Adjacent Normal")
legend("topright",legend=paste("mean = ", 
round(mean(foldchanges),2),
"\nStdev = ", round(sd(foldchanges),2)))
```

\newpage
##RnaXSim

###RNA-seq Raw Count Simulation
####Default


In this section, we will run RnaXSim with default settings as a demonstration.

1. Get partial RNA-seq data from SEQC Sample A.

```{r, echo=TRUE}
data(SEQC)
SampleA <- SEQC
#We are only using part of the matrix for an example.
#However, users are encouraged to use the whole matrix.
```


2. Simulate an RNA-seq dataset.
```{r, echo=TRUE}
library(Linnorm)
#This will generate two sets of RNA-seq data with 3 replicates each. 
#It will have 20000 genes totally with 5000 genes being differentially 
#expressed. It has the Poisson distribution.
SimulatedData <- RnaXSim(SampleA)
```


3. Separate data into matrices and vectors as an example.
```{r, echo=TRUE}
#Index of differentially expressed genes.
DE_Index <-